frontend/components/tree/js/tree.service.ts

```diff
@@ -1015,43 +1022,121 @@ export class TreeService {
 		this.ViewerService.clearHighlights();
 		this.DocsService.closeDocs();
 
-		if (this.currentSelectedNodes) {
-			for (let i = 0; i < this.currentSelectedNodes.length; i++) {
-				this.currentSelectedNodes[i].selected = false;
+		if (this.currentSelectedNodes.length) {
+			while (this.currentSelectedNodes.length) {
+				const currentNode = this.currentSelectedNodes.pop();
+				currentNode.selected = this.SELECTION_STATES.unselected;
+				this.setSelectionOnParentNodes(currentNode, this.SELECTION_STATES.unselected);
+				if (currentNode.children && currentNode.children.length) {
+					this.currentSelectedNodes = this.currentSelectedNodes.concat(currentNode.children);
+				}
 			}
 		}
 
 		this.currentSelectedNodes = [];
 	}
 
+	/**
+	 * Set the selection state of a nodes parents to represent it's children
+	 * @param nodes the nodes to set a new selection state on
+	 * @param select wether the node should be selected or not
+	 */
+	public traverseNodesAndSetSelected(nodes: any[], select: boolean) {
+
+		nodes = nodes.concat(); // make a copy
+		const setParentNodes = nodes.concat();
+		let currentNode;
+
+		while (nodes.length) {
+			currentNode = nodes.pop();
+
+			const nodeIndex = this.currentSelectedNodes.indexOf(currentNode);
+			const notParentOfUnselected = currentNode.selected !== this.SELECTION_STATES.parentOfUnselected;
+
+			if (select && notParentOfUnselected && nodeIndex === -1) {
+				this.currentSelectedNodes.push(currentNode);
+			} else if (nodeIndex > -1) {
+				this.currentSelectedNodes.splice(nodeIndex, 1);
+			}
+
+			if (currentNode.toggleState !== "invisible") {
+
+				if (currentNode.toggleState !== "parentOfInvisible" && select) {
+					currentNode.selected = this.SELECTION_STATES.selected;
+				} else if (!select) {
+					currentNode.selected = this.SELECTION_STATES.unselected;
+				} else {
+					currentNode.selected = this.SELECTION_STATES.parentOfUnselected;
+				}
+
+				if (currentNode.children && currentNode.children.length) {
+					nodes = nodes.concat(currentNode.children);
+				}
+			}
+
+		}
+
+		setParentNodes.forEach((node) => {
+			this.setSelectionOnParentNodes(node);
```

Comment:
Again, this is going through many nodes multiple times.
See comment on 1029.

This is worse since it's not a force set so it has to do a lot more logic as well!

Reply:
In reality this will only ever be of length 1

Reply:
if this will only be a length of 1 then I think you need to change your function definition, if it can be more than 1 then you need to do something similar to 1029.

Reply:
Mostly just thinking for extensibility but probably YAGNI